In [1]:
import numpy as np
from sympy import Matrix
import matplotlib.pyplot as plt
import seaborn as sns


In [5]:
A = np.array([[10, 3], [16, 4]])
#last digit 8 and 2nd last 2
size_A = A.shape

det_A = np.linalg.det(A)

if np.linalg.det(A) != 0:
    inv_A = np.linalg.inv(A)
else:
    inv_A = "Matrix is singular, cannot compute inverse"

rank_A = np.linalg.matrix_rank(A)

eigenvalues, eigenvectors = np.linalg.eig(A)

print("Size of A:", size_A)
print("Determinant of A:", det_A)
print("Inverse of A:\n", inv_A)
print("Rank of A:", rank_A)

print("Eigenvalues of A:", eigenvalues)
print("Eigenvectors of A:\n", eigenvectors)
print("Roots of the Characteristic Polynomial (Eigenvalues):", roots)

Size of A: (2, 2)
Determinant of A: -7.999999999999998
Inverse of A:
 [[-0.5    0.375]
 [ 2.    -1.25 ]]
Rank of A: 2
Eigenvalues of A: [14.54983444 -0.54983444]
Eigenvectors of A:
 [[ 0.55047268 -0.27352072]
 [ 0.83485318  0.96186611]]
Roots of the Characteristic Polynomial (Eigenvalues): [14.54983444 -0.54983444]


In [6]:
B = np.array([[10, 4], [16, 4]])
#change [10,3] to [10,4]
size_B = B.shape

det_B = np.linalg.det(B)

if np.linalg.det(B) != 0:
    inv_B = np.linalg.inv(B)
else:
    inv_B = "Matrix is singular, cannot compute inverse"

rank_B = np.linalg.matrix_rank(B)

eigenvalues, eigenvectors = np.linalg.eig(B)

print("Size of B:", size_B)
print("Determinant of B:", det_B)
print("Inverse of B:\n", inv_B)
print("Rank of B:", rank_B)

print("Eigenvalues of B:", eigenvalues)
print("Eigenvectors of B:\n", eigenvectors)
print("Roots of the Characteristic Polynomial (Eigenvalues):", roots)

Size of B: (2, 2)
Determinant of B: -23.999999999999993
Inverse of B:
 [[-0.16666667  0.16666667]
 [ 0.66666667 -0.41666667]]
Rank of B: 2
Eigenvalues of B: [15.54400375 -1.54400375]
Eigenvectors of B:
 [[ 0.5851057  -0.32740282]
 [ 0.81095704  0.94488486]]
Roots of the Characteristic Polynomial (Eigenvalues): [14.54983444 -0.54983444]


# 1.How did the determinant change and why?

 Think of the determinant for a $2 \times 2$ matrix as $ad - bc$ (top-left $\times$ bottom-right minus top-right $\times$ bottom-left).For $A$: $(10 \times 4) - (3 \times 16) = 40 - 48 = -8$For $B$: $(10 \times 4) - (4 \times 16) = 40 - 64 = -24$By increasing the top-right number from 3 to 4, you increased the subtraction part ($bc$), pulling the determinant further into the negative. Geometric-wise, the determinant represents the "area scaling factor" (and orientation flip, due to the negative sign) of the space transformed by the matrix. Stretching that component enlarged the area transformation factor by 3 times.

# 2. Did the rank change? Explain.

No, the rank stayed at 2.Why: Rank simply measures how many independent directions (or non-redundant rows/columns) a matrix has. For a $2 \times 2$ matrix, the rank only drops to 1 if one row is an exact multiple of the other, which makes the determinant 0. Since both determinants are non-zero ($-8$ and $-24$), both matrices keep full dimensional capability (rank 2). Neither row is a copy or simple scale of the other.
# 3. How did the eigenvalues respond to the value change?

The eigenvalues shifted from roughly [14.55, -0.55] in $A$ to [15.54, -1.54] in $B$.Why: Eigenvalues are the scale factors along special directional axes. Two quick checks tie them back to matrix properties:Sum (Trace): The sum of the diagonal numbers ($10 + 4 = 14$) must equal the sum of the eigenvalues. Notice that $14.55 + (-0.55) = 14$, and $15.54 + (-1.54) = 14$. The diagonal sum didn't change, so the sum of the eigenvalues couldn't change either!Product (Determinant): Multiplying the eigenvalues gives the determinant. Since the determinant dropped from $-8$ to $-24$, the two eigenvalues had to spread further apart—one growing more positive and the other growing more negative—to multiply out to $-24$ while keeping their sum at $14$.
# 4. Is B easier or harder to invert than A? Why?

Mathematically, they are equally easy.Why: Both are simple $2 \times 2$ matrices, so computing the inverse uses the exact same step-by-step formula.Numerically: $B$ is actually slightly better/more stable to invert than $A$. When a matrix's determinant gets very close to 0, it becomes "almost singular," making small computational errors blow up during inversion. Since $B$ has a determinant ($-24$) further away from 0 than $A$ ($-8$), $B$ is farther from being singular, making its inverse safer and less sensitive to rounding errors in computer arithmetic.